### PR TITLE
Modules for gathering facts about ZFS datasets and pools

### PR DESCRIPTION
**Pull request by xen0l**
##### ISSUE TYPE

- New Module Pull Request
##### COMPONENT NAME

zfs_facts.py
zpool_facts.py
##### ANSIBLE VERSION

```
ansible 2.1.2.0
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

These modules return facts about ZFS datasets and pools. They can be used for much better ZFS management with Ansible. They return ZFS dataset/pools properties as facts and based on those one can take decisions, e.g. delete snapshots older than X, report space usage, report property values, check ZFS pool health status etc.  These modules should run on all ZFS platforms and should work for both OpenZFS and ZFS implementations.

I also created a new category "zfs" under storage and move zfs.py there, so modules are kept together.

### CODE REVIEW

**Comment by xen0l** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/3075

**Comment by xen0l** [comment]
@ansible, I got this reviewed. Can this be merged?

**Comment by mattclay** [comment]
@xen0l The [tests](https://app.shippable.com/runs/584c7d7c3a097e0f00d7dcae/1/console) for this PR did not pass.

Try running `ansible-doc zfs zfs_facts zpool_facts` locally to see the failure.

**Comment by xen0l** [comment]
@mattclay sorry for the issue. I fixed it. It now passes checks.

File: lib/ansible/modules/storage/zfs/zfs_facts.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    name:
+        description:
+            - ZFS dataset name.
+        alias: [ "ds", "dataset" ]
+        type: str
+        required: yes
+    recurse:
+        description:
+            - Specifies if properties for any children should be recursively
+              display.
```
Comment (pyhalov) [inline lib/ansible/modules/storage/zfs/zfs_facts.py]: should be displayed

Comment (xen0l) [reply lib/ansible/modules/storage/zfs/zfs_facts.py]: Fixed.

File: lib/ansible/modules/storage/zfs/zfs_facts.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        cmd.append('-o')
+        cmd.append('name,property,value')
+        cmd.append(self.properties)
+        cmd.append(self.name)
+
+        (rc, out, err) = self.module.run_command(cmd)
+
+        if rc == 0:
+            for line in out.splitlines():
+                dataset, property, value = line.split('\t')
```
Comment (pyhalov) [inline lib/ansible/modules/storage/zfs/zfs_facts.py:231]: Does line parsing depend on presence on self.parsable? BTW, why could we need non-parsable output?

Comment (xen0l) [reply lib/ansible/modules/storage/zfs/zfs_facts.py:231]: No, parsable=True will return non-human friendly values (1K vs 1000) as _zfs get -p_ would.

Comment (pyhalov) [reply lib/ansible/modules/storage/zfs/zfs_facts.py:231]: OK

**Comment by xen0l** [comment]
ready_for_review

**Comment by xen0l** [comment]
@sivel you closed ansible/ansible-modules-extras#3075, which was reviewed there. This was reviewed by @jurajlutter there. Any chance to get this merged?

**Pull request merged**